File: problems_051to075/problem_074/solution.py
import os
import sys
from math import factorial

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = 0

        length_per_starter = {}

        for num in range(1, 1000000):
            target = num
            chain = []
            length_of_chain = None

            while True:
                if target in chain:
                    break

                chain += [target]
                target = sum(factorial(int(digit)) for digit in str(target))

                if target in length_per_starter:
                    length_of_chain = len(chain) + length_per_starter[target]
                    break

            if length_of_chain is None:
                length_of_chain = len(chain)

            length_per_starter[num] = length_of_chain

            if length_of_chain == 60:
                answer += 1

        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_051to075/problem_074/solution.js
const { sum } = require('mathjs');

const { factorial } = require('../../lib/toolbox');

const solution = () => {
  let answer = null;

  const lengthPerStarter = {};

  for (let num = 1; num < 1000000; num += 1) {
    let target = num;
    const chain = [];
    let lengthOfChain = null;

    while (!chain.includes(target)) {
      chain.push(target);
      target = sum(String(target).split('').map((digit) => Number(factorial(Number(digit)))));

      if (target in lengthPerStarter) {
        lengthOfChain = chain.length + lengthPerStarter[target];
        break;
      }
    }

    if (lengthOfChain === null) {
      lengthOfChain = chain.length;
    }

    lengthPerStarter[num] = lengthOfChain;

    if (lengthOfChain === 60) {
      answer += 1;
    }
  }

  return answer;
};

if (process.env.NODE_ENV !== 'test') {
  // eslint-disable-next-line no-console
  console.log(solution());
}

module.exports = solution;
